### ISSUE TITLE
Deprecate Thrift API

### ISSUE THREAD

**Issue by duarten** [scylladb/scylladb#3811]
We should deprecate the Thrift API with the goal of removing the code in some future version. The reasons for this are:

1. Past Scylla version have Thrift support, and work correctly;
2. We are unlikely to invest effort in making new features like SI and LWT work from Thrift;
3. Projects like KairosDB that supported Scylla as a backend and used Thrift have since moved to CQL;
4. Less code means faster compile times!

**Comment by glommer** [comment]
Disagree. There are some folks that are still using thrift, believe it or not. Some of them don't want to rewrite applications and would consider moving to Scylla instead of newer Cassandra.

We shouldn't deprecate things without very good reason. APIs are particularly problematic.

The "past scylla versions have thrift" is not a good argument in my view. Those past versions exist, but they are unsupported, so what good is that ?

We obviously don't have to invest more time in it, and should only fix critical bugs. But I see deprecation as the wrong thing to do.

**Comment by nyh** [comment]
@glommer please note that Cassandra dropped Thrift support in 4.0 (see https://issues.apache.org/jira/browse/CASSANDRA-11115) in a move that has been planned for two years with out too much opposition. So users who insist on continuing to using thrift would need to do more than "consider" to use Scylla instead of newer Cassandra - because newer Cassandra will not support it at all.
Personally, I think we should keep Thrift for a while longer, but mark it deprecated and definitely not recommended for new projects, and we don't need to support anything more in it than we already do.

**Comment by arganzheng** [comment]
Without Thrift API, how can we insert record without define every column before we can insert it.

**Comment by nyh** [comment]
After #8336, the Thrift protocol is no longer enabled by default. But it's still not officially deprecated.

**Comment by mykaul** [comment]
It's almost 2023, perhaps it's the year of deprecation of Thrift!

**Comment by DoronArazii** [comment]
@mykaul is it planned for 5.2?
Which team has the commitment?

**Comment by mykaul** [comment]
> @mykaul is it planned for 5.2? Which team has the commitment?

No critical, I assume @eliransin , could be postponed to 5.3.

**Comment by nyh** [comment]
Issues #1478 and #9911, which asked to improve testing of Thrift but never done - were closed recently.
Not having good tests is yet another reason to drop Thrift support.

**Comment by eliransin** [comment]
@mykaul , I am not sure about the decision. Should we deprecate thrift? If so, maybe in order to deprecate thrift in 5.3 we should at least start by issuing a warning that it is going to be deprecated in 5.2 and the actually deprecate it in 5.3?

**Comment by mykaul** [comment]
> @mykaul , I am not sure about the decision. Should we deprecate thrift? If so, maybe in order to deprecate thrift in 5.3 we should at least start by issuing a warning that it is going to be deprecated in 5.2 and the actually deprecate it in 5.3?

We can issue a warning, including in the release notes, in 5.2, and do it in 5.3.

**Comment by nyh** [comment]
I think that dropping Thrift is not really a technical decision - technically Thrift is bad and not needed and not tested enough (the issues to fix that we. I think it's a product/business decision:
* Do we know of users who use Thrift or want to use it? How many such users?
* Is our unique Thrift support (which Cassandra no longer supports since 4.0) attracting users to Scylla? 

@eliransin for two years now (since https://github.com/scylladb/scylladb/commit/ccc75bfe2ac2f1888f908e5cb1fb59def0f51313) Thrift is disabled in Scylla by default and needs to be enabled manually with the `start_rpc` configuration option. This means we can indeed warn users when Scylla starts with that option enable - that this option will be going away soon (?). The documentation (docs/dev/protocols.md) already un-recommends using Thrift for three years already:

> The Apache Thrift protocol was early Cassandra's client protocol, until it was superceded in Cassandra 1.2 with the binary CQL protocol. Thrift was still nominally supported by both Cassandra and Scylla for many years, but was recently dropped in Cassandra (version 4.0) and is likely to be dropped by Scylla in the future as well, so it is not recommended for new applications.

**Comment by avelanarius** [comment]
Additional point of view related to our tooling: scylla-tools-java contains code related to Thrift and uses a very old version of libthrift. That library is flagged by security scanners. Unfortunately, moving to a newer libthrift requires a lot of changes and Thrift testing (that we are lacking...). See https://github.com/scylladb/scylla-tools-java/pull/353 for more context, where I decided to effectively not support Thrift in scylla-tools-java no longer.

**Comment by kostja** [comment]
@mykaul why is this not assigned out? What needs to be done here?

**Comment by mykaul** [comment]
> @mykaul why is this not assigned out? What needs to be done here?

Delete all thrift related code at this point.
